WIP: Unify action representation

metamben opened the pull request:
[Dev doc](https://www.notion.so/metabase/Query-action-migration-039d06e147ec4c0eb98fd2c0177a6ef1)

- Make implicit actions first class
  - [x] Give them a DB table inheriting from action
  - [x] Get rid of `slug` when executing actions
  - [x] Update scaffolding
  - [x] Throw an exception when a model with actions is demoted to an ordinary query
- Bind actions directly to models
  - [x] Add `model_id` field
  - [x] Get rid of `model_action`
- Put common action properties into `action`
  - `parameters`
  - `parameter_mappings`
  - `visualization_settings`
- Migrations
  - [x] Add implicit_action  
  - [x] Put dataset_query on query_action
  - [x] Rollbacks
  - [x] Migrate existing actions
  - [x] Drop model_action
  - [x] Add not null constraints after existing migrations
  - [x] Remove card.is_write
- Make implicit action parameters customizable
  - [x] Updating parameters
- Permissions
  - [x] `action` API
- Serdes
- Cleanup
  - [x] Remove report_card.is_write
    - [x] card API
  - [x] Remove emitters

## Front end

Implemented in #26764

- [x] Save implicit actions with `{:type :implicit :kind <:row/create | :row/delete | :row/update> :visualization_settings ... :model_id ...}` to `POST /api/action`
- [x] Save query actions with `{:type :query :database_id ... :dataset_query ... :visualization_settings ... :parameters ... :model_id ...}` to `POST /api/action`  
- [x] Save dashcards with `action_id` AND `cardId` - Don't save `visualization_settings.action_slug`
- [x] Change execution/prefill paths to `dashboard/:id/dashcard/:id/execute`

snoe committed:
[Apps] WIP initial db migrations for giving implicit actions an action_id and removing card from query_action

snoe committed:
Make crud work for http actions

snoe committed:
Make implicit crud work

snoe committed:
Rename implicit_action.namespace to kind. Rewrite merged-model-actions to actions-with-implicit-parameters

metamben committed:
Fix query action creation

snoe committed:
Make actions.test-util/with-action create a holding model

metamben committed:
Fix unified-action-create-test

snoe committed:
Add tests for creating implicit and http dashcard actions

metamben committed:
Fix get-action-test

metamben committed:
Enable query action in dashcard-action-create-update-test

metamben committed:
Fix metabase.api.dashboard-test/dashcard-query-action-execution-test

metamben committed:
Fix metabase.api.dashboard-test/dashcard-http-action-execution-test

metamben committed:
Fix metabase.api.dashboard-test/dashcard-action-execution-auth-test

metamben committed:
Fix metabase.api.dashboard-test/dashcard-action-execution-granular-auth-test

metamben committed:
Merge remote-tracking branch 'origin/data-apps-main' into data-apps/phase-2-model-actions

snoe committed:
Make implicit action execution work

snoe committed:
Scaffolding implicit actions from table

metamben committed:
Fix metabase.api.action-test

metamben commented on src/metabase/actions/execution.clj:
```diff
@@ -177,18 +167,17 @@
-      (let [model-action (api/check-404 (first (action/actions-with-implicit-params nil :id (:id action))))]
-        (execute-implicit-action model-action request-parameters))
+        action (api/check-404 (first (action/actions-with-implicit-params nil :id (:action_id dashcard))))]
+    (if (= :implicit (:type action))
```
Cam told me this was Yoda speech when I wrote the condition this way. (Although we write tests like this (first comes the expected value).)

snoe replied:
Ah, for me it's left over java brain. `CONSTANT.equals(var)` won't throw an NPE.

metamben replied:
I might have the same in the subconscious.

snoe committed:
Make scaffolding work with existing models. Update with-action macro to allow multiple actions

iethree commented:
> Save dashcards with action_id AND cardId - Don't save visualization_settings.action_slug

There are a number of places on the frontend (mostly in the action entity) where we're doing non-ideal things with slugs since we didn't have unique ids from implicit actions. these can be cleaned up and simplified now! 🙌

metamben commented on src/metabase/api/app.clj:
```diff
@@ -288,11 +290,13 @@
+                  actions (->> actions
                                (filter selector)
                                (map action-mapper)
                                (sort-by (comp sorter first)))]]
```
Is `sorter` going to get a kind here? `action-mapper` seems to return an action-id as the first element and kind doesn't seem to be in the tuple at all.

metamben commented on test/metabase/actions/test_util.clj:
```diff
@@ -117,83 +117,89 @@
+         (mt/with-model-cleanup [Action]
+           (let [~@(mapcat (fn [[binding-form option-map]]
+                             [binding-form `(do-with-action (merge {:type :query} ~option-map) ~model-id)])
+                           (partition-all 2 binding-forms-and-option-maps))]
```
Would it make sense to insist on an even number of elements in `binding-forms-and-option-maps`? Only the last one can be omitted, otherwise you get a mess. Would you encourage `(with-action [{id :action-id}] ...)`?

snoe replied:
I think it makes sense for pairs otherwise I don't think it should be called a binding form. I would discourage that example as it feels very non-idiomatic. `(let [x nil] ...) vs (let [x] ...)`

metamben commented on test/metabase/actions/test_util.clj:
```diff
@@ -117,83 +117,89 @@
+                                        options-map))]
+        {:action-id action-id :model-id model-id}))))
 
 (defmacro with-action
```
Rename to `with-actions`?

metamben replied:
Done.

metamben committed:
Fix metabase.api.public-test

metamben committed:
Remove unused names from metabase.models.action

metamben committed:
Remove some unused imports

metamben committed:
Merge remote-tracking branch 'origin/data-apps-main' into data-apps/phase-2-model-actions

metamben committed:
Remove an unused import

metamben committed:
Remove model_action and fix a bunch of tests

metamben committed:
Fix dashcard-implicit-action-only-expose-and-allow-model-fields

metamben committed:
Fix action type inheritance

snoe committed:
Fill in remarks and add description to action

snoe committed:
Add forward migration for existing actions and rollbacks that remove existing actions

metamben committed:
Prevent overwriting parameters of custom actions

metamben committed:
Support setting description on actions

snoe committed:
Update migration to work with mysql - rollback is still not quite working in mysql due to database_id fk

metamben committed:
Rename with-action macro to with-actions and document it

metamben committed:
Remove report_card.is_write

metamben committed:
Fix some more tests

snoe committed:
Fix rollback across all appdb drivers, add not null constraints, use action.name to hold implicit slugs

snoe committed:
Pull name and description from is_write card

snoe committed:
Add name to scaffolded implicit actions

metamben committed:
Remove ModelAction

metamben committed:
Remove is_write from comments too

metamben committed:
Merge remote-tracking branch 'origin/data-apps-main' into data-apps/phase-2-model-actions

snoe committed:
Different databases reference capture groups in regexp_replace differently. Also drop model_action table

snoe committed:
Update migrations to pass linter.

Introduce ${capturegroup.ref} to handle the different ways to reference
capture groups within a regexp_replace. This allows collapsing large
data migrations into a single `sql` element that the linter requires.

snoe committed:
Add kind to action-mapper so we can sort as we wish

snoe committed:
Add name to with-temp Action

kulyk committed:
Merge branch 'data-apps-main' into data-apps/phase-2-model-actions

kulyk committed:
Adjust FE to query actions migration (#26764)

* Remove `is_write` card notion

* Remove action `slug` from action types

* Fix typo in `estimateCardSize` name

* Remove hanging card's `is_write` prop

* Remove not used writeback actions

* Update endpoints

* Remove not used prop

* Remove action `slug` usage

* Update actions entity

* Tweak action creator to handle new actions shape

* Fix saving action dashcards

* Fix actions API usage

How did that work before? 🤔

* Pass `database_id` when creating query actions

* Remove custom list API implementation for actions

* Clean up types

Co-authored-by: Tamás Benkő <tamas@metabase.com>

snoe committed:
Change regexp_replace to substring to support mysql - this is ok because we have consistent json serialization.

metamben committed:
Check permissions on the model in the action API

metamben committed:
Remove unused required namespace

metamben committed:
Throw an exception when demoting a model with actions to question

snoe committed:
Allow saved parameters on implicit actions to add to implicit parameters
